lms/djangoapps/grades/models.py

```diff
@@ -236,9 +241,43 @@ def bulk_get_or_create(cls, block_record_lists, course_key):
         BlockRecordList objects for the given course_key, but
         only for those that aren't already created.
         """
-        existent_records = {record.hashed: record for record in cls.bulk_read(course_key)}
+        existent_records = cls.bulk_read(course_key)
         non_existent_brls = {brl for brl in block_record_lists if brl.hash_value not in existent_records}
-        cls.bulk_create(non_existent_brls)
+        created = cls.bulk_create(non_existent_brls)
+        cls.update_cache(course_key, created)
```

Comment:
For a better separation of concerns, move the logic of updating the cache into the `bulk_create` method.  This way, the `bulk_get_or_create` method can simply be a higher layer method on top of the `bulk_read` and `bulk_create` methods and will read more easily.  And then the cache handling and updating can happen within the `bulk_read` and `bulk_create` methods.